### PR TITLE
Screenshare presence - sandbox automation for authorization post

### PR DESCRIPTION
**Pull request by howard-stearns**
https://highfidelity.atlassian.net/browse/DEV-2671
https://highfidelity.atlassian.net/wiki/spaces/PM/pages/540278806/Screenshare+Back-end+Support

This is do-not-merge until the backend endpoint is available (https://github.com/highfidelity/data-web/pull/2245, but it needs to be on production, too).

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Can someone from High Fidelity approve this PR?

**Comment by zfox23** [comment]
This looks great to me.

I made this:  https://highfidelity.atlassian.net/browse/DEV-2700

**Comment by zfox23** [comment]
Noting the DNM label on this PR!

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22352/HighFidelity-Beta-PR16459-d59f622.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22352/HighFidelity-Quest-Beta-PR16459-d59f622.apk)

File: assignment-client/src/avatars/AvatarMixerClientData.cpp [hunk 1]
```diff
@@ -152,7 +154,16 @@ int AvatarMixerClientData::parseData(ReceivedMessage& message, const SlaveShared
...
+        bool currentlyHasPriority = findPriorityZone.isInPriorityZone;
+        if (currentlyHasPriority != _avatar->getHasPriority()) {
+            _avatar->setHasPriority(currentlyHasPriority);
+            auto nodeList = DependencyManager::get<NodeList>();
+            const DomainHandler& domainHandler = nodeList->getDomainHandler();
+            auto packet = NLPacket::create(PacketType::AvatarZonePresence, 2 * NUM_BYTES_RFC4122_UUID, true);
+            packet->write(_avatar->getSessionUUID().toRfc4122());
+            packet->write(findPriorityZone.id.toRfc4122());
+            nodeList->sendPacket(std::move(packet), domainHandler.getSockAddr());
+        }
```
Comment (SimonWalton-HiFi) [inline assignment-client/src/avatars/AvatarMixerClientData.cpp]: Hijacking the hero zone is just for testing, right? Eventually the zone entity will have a new field for screen-share zones. We can combine the two zone searches in the one functor by adding another bool to it. I don't know if we need the full hero logic of turning on & off - probably nested zones are not part of the design.

Comment (howard-stearns) [reply assignment-client/src/avatars/AvatarMixerClientData.cpp]: indeed... See Zach's comments re https://highfidelity.atlassian.net/browse/DEV-2700

File: domain-server/src/DomainServer.cpp [hunk 2]
```diff
@@ -3613,3 +3614,62 @@ void DomainServer::handleOctreeFileReplacementRequest(QSharedPointer<ReceivedMes
...
+        return;
+    }
+    auto limitedNodeList = DependencyManager::get<LimitedNodeList>();
+    auto matchingNode = limitedNodeList->nodeWithUUID(avatar);
+    if (!matchingNode) {
+        qCWarning(domain_server) << "Ignoring avatar presence for unknown avatar" << avatar;
+        return;
+    }
+    QString verifiedUsername = matchingNode->getPermissions().getVerifiedUserName();
+    static int SCREENSHARE_EXPIRATION_SECONDS = 24 * 60 * 60;
```
Comment (SimonWalton-HiFi) [inline domain-server/src/DomainServer.cpp]: Could be `const` or `constexpr` for added efficiency.

File: domain-server/src/DomainServer.cpp [hunk 3]
```diff
@@ -3613,3 +3614,62 @@ void DomainServer::handleOctreeFileReplacementRequest(QSharedPointer<ReceivedMes
...
+void DomainServer::handleSuccessfulScreensharePresence(QNetworkReply* requestReply) {
+    QJsonObject jsonObject = QJsonDocument::fromJson(requestReply->readAll()).object();
+    if (jsonObject["status"].toString() != "success") {
+        qCWarning(domain_server) << "screensharePresence api call failed:" << QJsonDocument(jsonObject).toJson(QJsonDocument::Compact);
+    }
+}
+
+void DomainServer::handleFailedScreensharePresence(QNetworkReply* requestReply) {
+    qCWarning(domain_server) << "screensharePresence api call failed:" << requestReply->error();
+}
```
Comment (SimonWalton-HiFi) [inline domain-server/src/DomainServer.cpp]: Github is warning of no newline.

File: assignment-client/src/avatars/AvatarMixerClientData.cpp [hunk 4]
```diff
@@ -152,7 +154,16 @@ int AvatarMixerClientData::parseData(ReceivedMessage& message, const SlaveShared
         EntityTree& entityTree = *slaveSharedData.entityTree;
         FindPriorityZone findPriorityZone { newPosition } ;
         entityTree.recurseTreeWithOperation(&FindPriorityZone::operation, &findPriorityZone);
-        _avatar->setHasPriority(findPriorityZone.isInPriorityZone);
+        bool currentlyHasPriority = findPriorityZone.isInPriorityZone;
+        if (currentlyHasPriority != _avatar->getHasPriority()) {
+            _avatar->setHasPriority(currentlyHasPriority);
+            auto nodeList = DependencyManager::get<NodeList>();
+            const DomainHandler& domainHandler = nodeList->getDomainHandler();
```
Comment (SimonWalton-HiFi) [inline assignment-client/src/avatars/AvatarMixerClientData.cpp]: There's actually a `NodeList::getDomainSockAddr()`  for convenience.

File: domain-server/src/DomainServer.cpp [hunk 5]
```diff
@@ -3613,3 +3614,62 @@ void DomainServer::handleOctreeFileReplacementRequest(QSharedPointer<ReceivedMes
...
+    if (!matchingNode) {
+        qCWarning(domain_server) << "Ignoring avatar presence for unknown avatar" << avatar;
+        return;
+    }
+    QString verifiedUsername = matchingNode->getPermissions().getVerifiedUserName();
+    static int SCREENSHARE_EXPIRATION_SECONDS = 24 * 60 * 60;
+    screensharePresence(zone.isNull() ? "" : zone.toString(), verifiedUsername, SCREENSHARE_EXPIRATION_SECONDS);
+}
+
+void DomainServer::screensharePresence(QString roomname, QString username, int expiration_seconds) {
```
Comment (SimonWalton-HiFi) [inline domain-server/src/DomainServer.cpp]: `expiration_seconds` - Ruby-style identifier!

Comment (howard-stearns) [reply domain-server/src/DomainServer.cpp]: :-)

**Comment by howard-stearns** [comment]
Thanks, @SimonWalton-HiFi. Updated.

File: assignment-client/src/avatars/AvatarMixerClientData.cpp [hunk 6]
```diff
@@ -99,6 +99,7 @@ namespace {
         glm::vec3 position;
         bool isInPriorityZone { false };
         float zoneVolume { std::numeric_limits<float>::max() };
+        EntityItemID id;
```
Comment (SimonWalton-HiFi) [inline assignment-client/src/avatars/AvatarMixerClientData.cpp]: I think the mac & linux builds were complaining about id not being initialized.

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22358/HighFidelity-Beta-PR16459-0b03788.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22358/HighFidelity-Quest-Beta-PR16459-0b03788.apk)

**Comment by howard-stearns** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22363/HighFidelity-Beta-PR16459-0b03788.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22363/HighFidelity-Quest-Beta-PR16459-0b03788.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22363/HighFidelity-Beta-PR16459-0b03788.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22363/HighFidelity-Beta-Interface-PR16459-0b03788.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22386/HighFidelity-Beta-PR16459-74a322f.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22386/HighFidelity-Quest-Beta-PR16459-74a322f.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22386/HighFidelity-Beta-PR16459-74a322f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22386/HighFidelity-Beta-Interface-PR16459-74a322f.dmg

**Comment by howard-stearns** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22389/HighFidelity-Beta-PR16459-74a322f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22389/HighFidelity-Beta-Interface-PR16459-74a322f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22389/HighFidelity-Beta-PR16459-74a322f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22389/HighFidelity-Beta-Interface-PR16459-74a322f.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22389/HighFidelity-Beta-PR16459-74a322f.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22389/HighFidelity-Quest-Beta-PR16459-74a322f.apk)

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22519/HighFidelity-Beta-PR16459-da91924.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22519/HighFidelity-Quest-Beta-PR16459-da91924.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22519/HighFidelity-Beta-PR16459-da91924.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22519/HighFidelity-Beta-Interface-PR16459-da91924.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22519/HighFidelity-Beta-PR16459-da91924.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22519/HighFidelity-Beta-Interface-PR16459-da91924.exe

**Comment by howard-stearns** [comment]
closing in favor of https://github.com/highfidelity/hifi/pull/16429, which has this and all the other stuff in one throw.

**Pull request closed without merging**